### DISCUSSION TITLE
JavaScript/Typescript implementation

### DISCUSSION THREAD

**Discussion by fabiofranzini** [microsoft/Power-Fx#384] [Q&A] [answered]
Hi, do you have a public plan to add the JavaScript / Typescript implementation of Power FX language and runtime?
It would be fine to have the current C# implementation that compiles the functions to js

**Comment by MikeStall** [comment, marked as answer]
That's definitely on the roadmap. 
PowerApps consumes the PowerFx.Core package from this repo for parsing and binding. But instead of consuming the interpreter here, it has an alternative Power Fx backend (written in C#) that compiles to JScript.  _We would like to open source the JS implementation, but no ETA yet._ 

We also have a 3rd backend that compiles to SQL - it is used for formula columns. We'd also like to open source that.

**Reply by fabiofranzini** [reply]
Thanks for the response!
It would be nice to have a timeline, because my company, like others, would be interested in this solution for embedding PowerFX in their client-side web applications.

**Reply by MikeStall** [reply]
What about running via Blazor and Web Assembly? https://dotnet.microsoft.com/en-us/apps/aspnet/web-apps/blazor

**Reply by fabiofranzini** [reply]
Yes, this is a viable way to do that, but the cost of all runtimes could be higher than having only a JavaScript runtime?
I read in another thread that you plan to post a Blazor/Web Assembly based example, correct?

**Reply by MikeStall** [reply]
We do plan to publish a Blazor example, we just haven't gotten to it yet.  
We have C# hosts that definitely need a C# runtime; and JS / Browser hosts that want a JS runtime; and we need SQL too - so we'll always have at least 3 different runtimes. 
This was part of the reason that we created .txt tests - they're agnostic to language and can be run against any runtime:
https://github.com/microsoft/Power-Fx/tree/main/src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases

**Reply by YousefED** [reply]
Hi @MikeStall ! Is this still on the roadmap? Would love to embed PowerFx in a javascript app!

**Reply by cchannon** [reply]
I'll toss in a vote for the same. While having a server context like Blazor to interact with is a technically viable option, it is also not very "Power Platform." An approach like that can't be deployed in a solution, it introduces release and configuration management headaches, and it can mean a lot of overhead to deploy securely, whereas if the eval happens only client-side those concerns pretty much evaporate.

**Reply by fabiofranzini** [reply]
Yes @cchannon but you can use Blazor not in the server context but into the client context.
This, while feasible, does not appeal to me because it still requires loading the Blazor runtime in addition to the PowerFx runtime. The best thing would be to have PowerFx's functions and runtime compiled directly into javascript

**Reply by MikeStall** [reply]
This is still on the roadmap, but no ETA. 

For now, your best bet is still blazor / wasm - there is an example here: https://github.com/microsoft/Power-Fx/discussions/140
PVA just went into public preview using wasm. 

The goal is that Power Fx is fully bundled with Power Platform, so you don't have to deploy it. For example, Power Fx is already being used in UCI and forms for Commanding - we don't want to load a 2nd JScript Power Fx runtime for the solution.

**Reply by YousefED** [reply]
Thanks for the quick replies. My use-case is to use it as a scripting system outside of the Power Platform. The wasm implementation is interesting, but not really feasible (20mb, etc). Looking forward to the JS / TS version!

Is Power Platform internally using a Javascript version? Or a cross-compiled / wasm solution?

**Comment by filcole** [comment]
I will write a blog on how to do this by using a Custom API in the Power Platform.  It's still a bit of hack until it's fully deployed (and uses preview features).

**Reply by BetimBeja** [reply]
if you mean the blogpost from @filcole you can find it here https://philcole.org/post/powerfx-customapi/

**Reply by rogerhendriks** [reply]
Sorry wrong thread :)

**Comment by rogerhendriks** [comment]
@MikeStall hi, any update on the JavaScript / TypeScript status?

**Reply by MikeStall** [reply]
sorry, not yet. Primary focus has been on C# interpreter to enable PVA, Cards, Datavverse Plugins  (more at #1529)
